crates/pacdef_core/src/backend/actual/rustup.rs

```diff
@@ -0,0 +1,208 @@
+use crate::backend::backend_trait::{Backend, Switches, Text};
+use crate::backend::macros::impl_backend_constants;
+use crate::{Group, Package};
+use anyhow::Context;
+use core::panic;
+use std::collections::HashSet;
+use std::os::unix::process::ExitStatusExt;
+use std::process::Command;
+
+#[derive(Debug, Clone)]
+pub struct Rustup {
+    pub(crate) packages: HashSet<Package>,
+}
+
+const BINARY: Text = "rustup";
+const SECTION: Text = "rustup";
+
+const SWITCHES_INSTALL: Switches = &["component", "add"];
+const SWITCHES_INFO: Switches = &["component", "list", "--installed"];
+const SWITCHES_MAKE_DEPENDENCY: Switches = &[];
+const SWITCHES_NOCONFIRM: Switches = &[];
+const SWITCHES_REMOVE: Switches = &["component", "remove"];
+
+const SUPPORTS_AS_DEPENDENCY: bool = false;
+
+impl Backend for Rustup {
+    impl_backend_constants!();
+
+    fn get_all_installed_packages(&self) -> anyhow::Result<HashSet<Package>> {
+        let mut toolchains_vec = self
+            .run_toolchain_command(self.get_info_switches("toolchain"))
+            .context("Getting installed toolchains")?;
+
+        let mut toolchains: HashSet<Package> = toolchains_vec
+            .iter()
+            .map(|name| ["toolchain", name].join("/").into())
+            .collect();
+
+        let packages: HashSet<Package> = self
+            .run_component_command(self.get_info_switches("component"), &mut toolchains_vec)
+            .context("Getting installed components")?
+            .iter()
+            .map(|name| ["component", name].join("/").into())
+            .collect();
+        toolchains.extend(packages.into_iter());
+        Ok(toolchains)
+    }
+
+    fn get_explicitly_installed_packages(&self) -> anyhow::Result<HashSet<Package>> {
+        self.get_all_installed_packages()
+            .context("Getting all installed packages")
+    }
+
+    fn make_dependency(&self, _: &[Package]) -> anyhow::Result<std::process::ExitStatus> {
+        panic!("Not supported by {}", self.get_binary())
+    }
+
+    fn install_packages(
+        &self,
+        packages: &[Package],
+        _: bool,
+    ) -> anyhow::Result<std::process::ExitStatus> {
+        let mut result: anyhow::Result<std::process::ExitStatus> =
+            Ok(std::process::ExitStatus::from_raw(0));
+        for p in packages {
+            let repo = p
+                .repo
+                .as_ref()
+                .expect("Not specified whether it is a toolchain or a component!");
```

Comment:
I'm slightly confused on this. Where all should I panic out and where should I not?

Reply:
In Rust, panics are generally for programming errors and unrecoverable errors. Result types (and their Err variant) are used for user input and recoverable errors. 

For example:
- You are using a library. Some function must be called in a specific way. The function checks whether this is done. It finds that the call is not correct. So there is obviously a bug in how the API is used. The function panics the program so that the programmer can fix it before deploying it to the user. 
- unrecoverable errors are things like OOM, and basic things breaking in unforeseeable ways. Not being able to write to disk is not one of them, because that can happen for plenty of reasons. (Permission, disk is full, ...)

Recoverable errors:
- A program works with user input. User input can be all kinds of wrong, so it must be checked whether for correctness and well-formedness. In case it isn't correct, you want to provide a proper, user-friendly error message that can be understood by just about anyone. Ideally with a nice backtrace for debugging purposes. **That's what pacdef does and applies to your question.**
- A web server uses worker threads. For some reason one of the worker threads dies. You may want to panic the server program, but you could also argue that the impact would be less severe if you start a new thread to replace the old one, and let the user request the webpage again. So you have a recoverable error. 

Does that make sense?

Reply:
Thanks, that helped clear stuff up. I mainly wanted to have it as an unrecoverable error as we depended on the user input being right and passing that input to an external program but what you have said makes more sense.